tests/cpptests/test_cpp_resultprocessor.cpp

```diff
@@ -88,3 +104,981 @@ TEST_F(ResultProcessorTest, testProcessorChain) {
   ASSERT_EQ(2, numFreed);
   RLookup_Cleanup(&lk);
 }
+
+TEST_F(ResultProcessorTest, testHybridMergerSameDocs) {
+  QueryIterator qitr = {0};
+
+  // Mock upstream1: generates 3 docs with score 2.0 (e.g., text search results)
+  struct MockUpstream1 : public ResultProcessor {
+    int counter = 0;
+    static int NextFn(ResultProcessor *rp, SearchResult *res) {
+      MockUpstream1 *p = static_cast<MockUpstream1 *>(rp);
+      if (p->counter >= 3) return RS_RESULT_EOF;
+
+      res->docId = ++p->counter;
+      res->score = 2.0;
+
+      // Set up document metadata with keys
+      static RSDocumentMetadata dmd[3] = {0};
+      static const char* keys[] = {"doc1", "doc2", "doc3"};
+      dmd[p->counter - 1].keyPtr = (char*)keys[p->counter - 1];
+      res->dmd = &dmd[p->counter - 1];
+
+      return RS_RESULT_OK;
+    }
+    MockUpstream1() {
+      memset(this, 0, sizeof(*this));
+      this->Next = NextFn;
+    }
+  } upstream1;
+
+  // Mock upstream2: generates same 3 docs with score 4.0 (e.g., vector search results)
+  struct MockUpstream2 : public ResultProcessor {
+    int counter = 0;
+    static int NextFn(ResultProcessor *rp, SearchResult *res) {
+      MockUpstream2 *p = static_cast<MockUpstream2 *>(rp);
+      if (p->counter >= 3) return RS_RESULT_EOF;
+
+      res->docId = ++p->counter; // Same docIds as upstream1
+      res->score = 4.0;
+
+      // Set up document metadata with keys (same docs as upstream1)
+      static RSDocumentMetadata dmd[3] = {0};
+      static const char* keys[] = {"doc1", "doc2", "doc3"};
+      dmd[p->counter - 1].keyPtr = (char*)keys[p->counter - 1];
+      res->dmd = &dmd[p->counter - 1];
+
+      return RS_RESULT_OK;
+    }
+    MockUpstream2() {
+      memset(this, 0, sizeof(*this));
+      this->Next = NextFn;
+    }
+  } upstream2;
+
+
+
+  // Create hybrid merger with window size 4
+  ResultProcessor *upstreams[] = {&upstream1, &upstream2};
+  ResultProcessor *hybridMerger = RPHybridMerger_New(
+    hybridScoringFunction,
+    upstreams,
+    2,  // numUpstreams
```

Comment:
The number of upstreams is hardcoded in a macro, and also an argument to the merger constructor - let's have it only an argument